src/xml_parser_main_metadata_together.c

```diff
@@ -336,132 +326,213 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
+    new_iter->filelists_pd = filelists_parser_data_new(newpkgcb_filelists, cbdata, pkgcb_filelists, cbdata, warningcb, warningcb_data);
+    new_iter->other_pd = other_parser_data_new(newpkgcb_other, cbdata, pkgcb_other, cbdata, warningcb, warningcb_data);
+    return new_iter;
+}
+
+int cr_PkgIterator_parseNext(cr_PkgIterator *iter, cr_Package **package, GError **err) {
+    cr_CbData *cbdata = (cr_CbData*) &iter->cbdata;
+    int ret = CRE_OK;
+
+    while (!cr_PkgIterator_is_finished(iter) && !cbdata->finished_pkgs_list) {
+        while ((cbdata->in_progress_count_primary <= cbdata->in_progress_count_filelists ||
+                cbdata->in_progress_count_primary <= cbdata->in_progress_count_other) &&
+               !iter->primary_is_done)
         {
-            primary_is_done = parse_next_section(primary_f, primary_path, primary_pd, err);
+            iter->primary_is_done = parse_next_section(iter->primary_f, iter->primary_path, iter->primary_pd, err);
             if (*err) {
                 ret = (*err)->code;
-                goto out;
+                cr_PkgIterator_free(iter, err);
             }
-
         }
 
-        while ((cbdata.in_progress_count_filelists <= cbdata.in_progress_count_primary ||
-                cbdata.in_progress_count_filelists <= cbdata.in_progress_count_other) &&
-               !filelists_is_done)
+        while ((cbdata->in_progress_count_filelists <= cbdata->in_progress_count_primary ||
+                cbdata->in_progress_count_filelists <= cbdata->in_progress_count_other) &&
+               !iter->filelists_is_done)
         {
-            filelists_is_done = parse_next_section(filelists_f, filelists_path, filelists_pd, err);
+            iter->filelists_is_done = parse_next_section(iter->filelists_f, iter->filelists_path, iter->filelists_pd, err);
             if (*err) {
                 ret = (*err)->code;
-                goto out;
+                cr_PkgIterator_free(iter, err);
             }
         }
 
-        while ((cbdata.in_progress_count_other <= cbdata.in_progress_count_filelists ||
-                cbdata.in_progress_count_other <= cbdata.in_progress_count_primary) &&
-               !other_is_done)
+        while ((cbdata->in_progress_count_other <= cbdata->in_progress_count_filelists ||
+                cbdata->in_progress_count_other <= cbdata->in_progress_count_primary) &&
+               !iter->other_is_done)
         {
-            other_is_done = parse_next_section(other_f, other_path, other_pd, err);
+            iter->other_is_done = parse_next_section(iter->other_f, iter->other_path, iter->other_pd, err);
             if (*err) {
                 ret = (*err)->code;
-                goto out;
+                cr_PkgIterator_free(iter, err);
             }
         }
     }
 
-out:
-    if (ret != CRE_OK) {
+    *package = g_slist_nth_data(cbdata->finished_pkgs_list, 0);
+    if (*package) {
+        g_slist_remove(cbdata->finished_pkgs_list, 0);
```

Comment:
I think `g_slist_remove` take a pointer not a position.
Perhaps this should work better:
```suggestion
        cbdata->finished_pkgs_list = cbdata->finished_pkgs_list->next;
```